### PR TITLE
Ymir imaging I/O helpers

### PR DESCRIPTION
**Pull request by YooSunYoung**
Related to #34 

Reference: https://github.com/scipp/ess-legacy/blob/683747138f1775d106809bf17bf5f56b75d6e0d3/imaging/helper_funcs.py#L41-L68

TODO
- [x] I think open beam and dark current are the other way around. 
- [x] Add test data and tests
- [x] Add documentation
- [ ] Add integration test ?

### CODE REVIEW

File: src/ess/imaging/io.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
+
+from collections.abc import Callable, Iterable
+from enum import Enum
+from itertools import pairwise
+from pathlib import Path
+from typing import NewType
+
+import jax.numpy as np
```
Comment (SimonHeybrock) [inline src/ess/imaging/io.py:10]: Missing in project dependencies, or is this unintentional?

File: src/ess/imaging/io.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    image_detector_name: ImageDetectorName,
+    histogram_mode_detectors_path: HistogramModeDetectorsPath = DEFAULT_HISTOGRAM_PATH,
+) -> HistogramModeDetectorData:
+    with File(file_path, mode="r") as f:
+        detector_dataset = f[
+            f"{histogram_mode_detectors_path}/{image_detector_name}/data"
+        ]
+        x_length, y_length = detector_dataset["value"].shape[1:]
+        return HistogramModeDetectorData(
+            sc.sort(
```
Comment (SimonHeybrock) [inline src/ess/imaging/io.py:56]: I think such a sort can be quite costly. Is this required?

File: src/ess/imaging/io.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    OPEN_BEAM = 2
+
+
+def load_nexus_histogram_mode_detector_data(
+    *,
+    file_path: FilePath,
+    image_detector_name: ImageDetectorName,
+    histogram_mode_detectors_path: HistogramModeDetectorsPath = DEFAULT_HISTOGRAM_PATH,
+) -> HistogramModeDetectorData:
+    with File(file_path, mode="r") as f:
```
Comment (SimonHeybrock) [inline src/ess/imaging/io.py:50]: Please use `h5py.File` for clarity which `File` this is.

If this is NeXus, why can't we use ScippNexus?

**Review by SimonHeybrock — COMMENTED** [review]
Is this for some non-compliant Nexus files, or why can't we use ScippNexus?

**Comment by YooSunYoung** [comment]
> Is this for some non-compliant Nexus files, or why can't we use ScippNexus?

The 2d detector images are stored in `int16`, it crashes with scippnexus.
Or is there a way you can skip incompatible part?

File: src/ess/imaging/io.py (hunk 1 above)
Comment (YooSunYoung) [reply src/ess/imaging/io.py]: Yes it was unintentional...

File: src/ess/imaging/io.py (hunk 2 above)
Comment (YooSunYoung) [reply src/ess/imaging/io.py:57]: We need to slice the data based on the time so it should be sorted along the time.
I think the data is already almost sorted in most cases.

**Review by YooSunYoung — COMMENTED** [review]
I used scippnexus wherever I could

**Comment by YooSunYoung** [comment]
Hmm...! I was wrong. I thought scippnexus raised an error when loading the images but it seems okay. I replaced the h5py to scippnexus.

**Comment by YooSunYoung** [comment]
We decided not to use data array to separate images, but the dictionary.
I'll reopen it when it's fixed.

File: src/ess/imaging/io.py [hunk 4]
```diff
@@ -93,8 +97,62 @@ def load_nexus_histogram_mode_detector(
...
+    max_coord: sc.Variable | None,
+) -> sc.DataArray:
+    if min_coord is not None and max_coord is not None:
+        return da[dim, min_coord:max_coord]
+    elif min_coord is not None and max_coord is None:
+        return da[dim, min_coord:]
+    elif min_coord is None and max_coord is not None:
+        return da[dim, :max_coord]
+    else:
+        return da
```
Comment (jl-wynen) [inline src/ess/imaging/io.py:136]:
```suggestion
    return da[dim, min_coord:max_coord]
```
You are reimplementing the behaviour of `slice` here.

Comment (YooSunYoung) [reply src/ess/imaging/io.py:136]: Ah... so I don't have to handle `None`...? That's great.

Comment (YooSunYoung) [reply src/ess/imaging/io.py]: I fixed it!

File: src/ess/imaging/io.py [hunk 5]
```diff
@@ -93,8 +97,62 @@ def load_nexus_histogram_mode_detector(
...
+"""Maximum value of the first dimension."""
+MinDim2 = NewType("MinDim2", sc.Variable | None)
+"""Minimum value of the second dimension."""
+MaxDim2 = NewType("MaxDim2", sc.Variable | None)
+
+
+def _assign_coord_or_make_one(
+    da: sc.DataArray, coord_name: str, coord: sc.Variable | None
+) -> None:
+    da.coords[coord_name] = coord or sc.arange(
```
[the remark this answers is not in the document]
Comment (YooSunYoung) [reply src/ess/imaging/io.py]: I fixed it...! Now it makes a dimensionless coordinate only if there is no coordinate.

File: src/ess/imaging/io.py [hunk 6]
```diff
@@ -113,27 +113,16 @@ def load_nexus_histogram_mode_detector(
...
-    if min_coord is not None and max_coord is not None:
-        return da[dim, min_coord:max_coord]
-    elif min_coord is not None and max_coord is None:
-        return da[dim, min_coord:]
-    elif min_coord is None and max_coord is not None:
-        return da[dim, :max_coord]
+    if (dim in da.coords.keys()) and (coord is not None):
+        da.coords[dim] = coord
+    elif (dim not in da.coords.keys()) and (coord is None):
+        da.coords[dim] = sc.arange(dim=dim, start=0, stop=da.sizes[dim])
```
Comment (jl-wynen) [inline src/ess/imaging/io.py:121]: This changed the logic. Now it raises if `(dim not in da.coords) and (coord is not None)`. Previously, it would insert the new coord.

Comment (YooSunYoung) [reply src/ess/imaging/io.py:121]: Yes, we don't want a new coordinate if it already exists.
I just wanted to make sure it has a coordinate before slicing.
So, if the loaded data already has a coordinate, you can change the cropping range according to that.

Comment (jl-wynen) [reply src/ess/imaging/io.py:121]: But that contradicts the code. If a coord exists and `coord is not None`, you overwrite the existing coord. The case I pointed out above would insert a coord if there is non already.

Comment (YooSunYoung) [reply src/ess/imaging/io.py:121]: Yeah I added that when we want to insert the coordinate but it should be in the loader function not in here.. I'll remove that part....

Comment (YooSunYoung) [reply src/ess/imaging/io.py:111]: Done...!

File: src/ess/imaging/io.py [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+    da: sc.DataArray, log: sc.DataArray, out_of_range_value: sc.Variable
+) -> sc.Variable:
+    """Sort the logs by time and decide which log entry corresponds to each time bin.
+
+    It assumes a log value is valid until the next log entry.
+    """
+    sorted_logs = sc.sort(log, 'time')
+    min_time = sc.datetime(da.coords['time'].min().value - 1, unit='ns')
+    max_time = sc.datetime(da.coords['time'].max().value + 1, unit='ns')
+    time_edges = sc.concat((min_time, sorted_logs.coords['time'], max_time), 'time')
```
[the remark this answers is not in the document]
Comment (YooSunYoung) [reply src/ess/imaging/io.py]: It was because of the out-or-range-values and assumming that logs are always less than the actual time coordinate but I think it can go wrong pretty easily... 

I updated it so that it just slice the dataset according to the log.

**Review by jl-wynen — APPROVED** [review]
